### ISSUE TITLE
too many concurrent call on the same entity

### ISSUE THREAD

**Issue by alirf81** [LuatixHQ/opencti#3019]
## Description

I am running 3 workers to import events from MISP.
While it processes a bundle of more than 12M operations, one of worker is not working with the error of `too many concurrent call on the same entity`

## Environment

1. OS (where OpenCTI server runs): Ubuntu 22.04 with docker-compose
2. OpenCTI version: OpenCTI 5.6.2
3. OpenCTI client: 
4. Other environment details: 

## Screenshots (optional)
![image](https://user-images.githubusercontent.com/92157864/225236685-a20f0c8f-16f0-4d2b-8ad9-df74c6aab87a.png)
![image](https://user-images.githubusercontent.com/92157864/225236743-97fd2963-04a8-4877-99e3-56b8623258a7.png)

**Comment by alirf81** [comment]
![image](https://user-images.githubusercontent.com/92157864/225274224-b02fefef-7122-4d88-9659-5493bd5a8da1.png)
After a while, it showed the above traceback and continued working.
Not sure what's going on ...

**Comment by richard-julien** [comment]
Hi, its a contention due to data/capacity of your deployments.
Could be limited by reducing the number of workers or improving your elasticsearch cluster speed.

**Comment by michaelheider** [comment]
I'm facing the same problem.

## TL;DR

I have three workers. 2 deadlock, the third one makes no progress.

Interestingly, even with just 1 worker, I still get the same problem. This speaks somewhat against the deadlock assumption and potentially hints at some other problem with not releasing a lock (?).

## What I tried

- Restart. Several times. Also, reboot. Deadlocks almost immediately again.
- Increase number of workers to 5. The first time I started the platform like this, 2 deadlocked right away, the rest made progress. But eventually, more workers joined the deadlock. And the system was dead again. Restarting, again with 5 workers, but now they all 5 go into deadlock very quickly. Similar results with 15 workers.
- Manually restart deadlocked workers. This doesn't help, they go back to the deadlock.
- Decrease to 1 worker. Which, in theory, would mean no deadlocks. However, the one worker still gets `Execution timeout, too many concurrent call on the same entities`. Which is weird.
- More memory to Elasticsearch. 20GB total did not help.

### What Helped

Increasing Elasticsearch's memory to 32GB and launching 150 workers (!) helped. Some of the workers (~6 before I let it run overnight) deadlocked again very quickly, but enough kep running to keep going and grind through the data to be imported (~168K queued bundles). After all bundles were completed, the error message stopped appearing completely.

One more thing was different: Over night, my workstation's connection to the server dropped, meaning this was the only attempt where I did not have some browser tabs, most of the time a sample report (the report in question was not tagged with TLP:GREEN), open in OpenCTI's web GUI. Not sure how this would make a difference, but I thought I'd mention it.

My setup is now back to a reasonable amount of workers.

## Log

Note how it always deadlocks on `marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da`, which is the marking definition 'TLP:GREEN'. This is consistent over multiple runs with different settings.

<details>
<summary>Log with 3 workers</summary>

Generated with OpenCTI 5.9.0 with docker compose on Ubuntu 20.04. 3 workers. Two deadlock, the third one keeps doing `{"timestamp": "2023-07-12T13:35:42.468812Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}`.

Only relevant logs. This is not the default log format, but made to be readable.

```json
opencti-worker-2
{
    "timestamp": "2023-07-12T12:55:30.554682Z",
    "level": "INFO",
    "name": "pycti.api",
    "message": "Importing a marking-definition (id: marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da)"
}

opencti-worker-3
{
    "timestamp": "2023-07-12T12:55:37.760695Z",
    "level": "INFO",
    "name": "pycti.api",
    "message": "Importing a marking-definition (id: marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da)"
}

opencti
{
    "category": "APP",
    "error": {
        "data": {
            "category": "business",
            "http_status": 500,
            "participantIds": [
                "marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da"
            ],
            "reason": "Execution timeout, too many concurrent call on the same entities"
        },
        "stacktrace": [
            "LockError: Lock timeout",
            "at error (/opt/opencti/build/src/config/errors.js:8:10)",
            "at LockTimeoutError (/opt/opencti/build/src/config/errors.js:100:51)",
            "at createEntityRaw (/opt/opencti/build/src/database/middleware.js:3260:13)",
            "at processTicksAndRejections (node:internal/process/task_queues:95:5)",
            "at createEntity (/opt/opencti/build/src/database/middleware.js:3285:16)",
            "at addMarkingDefinition (/opt/opencti/build/src/domain/markingDefinition.js:24:19)"
        ]
    },
    "inner_relation_creation": 0,
    "level": "warn",
    "message": "API Call",
    "operation": "MarkingDefinitionAdd",
    "operation_query": "mutation MarkingDefinitionAdd($input:MarkingDefinitionAddInput!){markingDefinitionAdd(input:$input){id standard_id entity_type parent_types definition_type definition x_opencti_order x_opencti_color created modified created_at updated_at}}",
    "size": 268,
    "time": 26206,
    "timestamp": "2023-07-12T12:55:56.779Z",
    "type": "WRITE_ERROR",
    "user": {
        "applicant_id": "88ec0c6a-13ce-5e39-b486-354fe4a7084f",
        "call_retry_number": "5",
        "group_ids": [
            "bbd1be92-4081-44bd-b8ed-b0c8733b9703"
        ],
        "ip": "::ffff:172.25.0.15",
        "organization_ids": [],
        "socket": "query",
        "user_id": "88ec0c6a-13ce-5e39-b486-354fe4a7084f"
    },
    "version": "5.9.0"
}

opencti-worker-2
{
    "timestamp": "2023-07-12T12:55:56.785660Z",
    "level": "ERROR",
    "name": "pycti.api",
    "message": "Execution timeout, too many concurrent call on the same entities"
}

opencti
{
    "category": "APP",
    "error": {
        "data": {
            "category": "business",
            "http_status": 500,
            "participantIds": [
                "marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da"
            ],
            "reason": "Execution timeout, too many concurrent call on the same entities"
        },
        "stacktrace": [
            "LockError: Lock timeout",
            "at error (/opt/opencti/build/src/config/errors.js:8:10)",
            "at LockTimeoutError (/opt/opencti/build/src/config/errors.js:100:51)",
            "at createEntityRaw (/opt/opencti/build/src/database/middleware.js:3260:13)",
            "at processTicksAndRejections (node:internal/process/task_queues:95:5)",
            "at createEntity (/opt/opencti/build/src/database/middleware.js:3285:16)",
            "at addMarkingDefinition (/opt/opencti/build/src/domain/markingDefinition.js:24:19)"
        ]
    },
    "inner_relation_creation": 0,
    "level": "warn",
    "message": "API Call",
    "operation": "MarkingDefinitionAdd",
    "operation_query": "mutation MarkingDefinitionAdd($input:MarkingDefinitionAddInput!){markingDefinitionAdd(input:$input){id standard_id entity_type parent_types definition_type definition x_opencti_order x_opencti_color created modified created_at updated_at}}",
    "size": 268,
    "time": 26076,
    "timestamp": "2023-07-12T12:56:03.843Z",
    "type": "WRITE_ERROR",
    "user": {
        "applicant_id": "88ec0c6a-13ce-5e39-b486-354fe4a7084f",
        "call_retry_number": "6",
        "group_ids": [
            "bbd1be92-4081-44bd-b8ed-b0c8733b9703"
        ],
        "ip": "::ffff:172.25.0.17",
        "organization_ids": [],
        "socket": "query",
        "user_id": "88ec0c6a-13ce-5e39-b486-354fe4a7084f"
    },
    "version": "5.9.0"
}

opencti-worker-3
{
    "timestamp": "2023-07-12T12:56:03.846400Z",
    "level": "ERROR",
    "name": "pycti.api",
    "message": "Execution timeout, too many concurrent call on the same entities"
}
```

</details>

<details>
<summary>Log with 1 worker</summary>

Generated with OpenCTI 5.9.0 with docker compose on Ubuntu 20.04. 1 worker. Only the logs of the worker. Repeating lines cut out.

```json
{"timestamp": "2023-07-12T13:35:34.265832Z", "level": "ERROR", "name": "pycti.api", "message": "A connection error occurred: {{ Traceback (most recent call last):
File \"/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py\", line 790, in urlopen
  response = self._make_request(
             ^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py\", line 536, in _make_request
  response = conn.getresponse()
             ^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/urllib3/connection.py\", line 454, in getresponse
  httplib_response = super().getresponse()
                     ^^^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/http/client.py\", line 1378, in getresponse
  response.begin()
File \"/usr/local/lib/python3.11/http/client.py\", line 318, in begin
  version, status, reason = self._read_status()
                            ^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/http/client.py\", line 287, in _read_status
  raise RemoteDisconnected(\"Remote end closed connection without\"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File \"/usr/local/lib/python3.11/site-packages/requests/adapters.py\", line 486, in send
  resp = conn.urlopen(
         ^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py\", line 844, in urlopen
  retries = retries.increment(
            ^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/urllib3/util/retry.py\", line 470, in increment
  raise reraise(type(error), error, _stacktrace)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/urllib3/util/util.py\", line 38, in reraise
  raise value.with_traceback(tb)
File \"/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py\", line 790, in urlopen
  response = self._make_request(
             ^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py\", line 536, in _make_request
  response = conn.getresponse()
             ^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/urllib3/connection.py\", line 454, in getresponse
  httplib_response = super().getresponse()
                     ^^^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/http/client.py\", line 1378, in getresponse
  response.begin()
File \"/usr/local/lib/python3.11/http/client.py\", line 318, in begin
  version, status, reason = self._read_status()
                            ^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/http/client.py\", line 287, in _read_status
  raise RemoteDisconnected(\"Remote end closed connection without\"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File \"/opt/opencti-worker/worker.py\", line 224, in data_handler
  self.api.stix2.import_bundle_from_json(
File \"/usr/local/lib/python3.11/site-packages/pycti/utils/opencti_stix2.py\", line 209, in import_bundle_from_json
  return self.import_bundle(
         ^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/pycti/utils/opencti_stix2.py\", line 2357, in import_bundle
  self.import_object(item, update, types)
File \"/usr/local/lib/python3.11/site-packages/pycti/utils/opencti_stix2.py\", line 814, in import_object
  stix_object_results = do_import(
                        ^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/pycti/entities/opencti_marking_definition.py\", line 298, in import_from_stix2
  return self.opencti.marking_definition.create(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/pycti/entities/opencti_marking_definition.py\", line 165, in create
  result = self.opencti.query(
           ^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/pycti/api/opencti_api_client.py\", line 333, in query
  r = self.session.post(
      ^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/requests/sessions.py\", line 637, in post
  return self.request(\"POST\", url, data=data, json=json, **kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/requests/sessions.py\", line 589, in request
  resp = self.send(prep, **send_kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/requests/sessions.py\", line 703, in send
  r = adapter.send(request, **kwargs)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/usr/local/lib/python3.11/site-packages/requests/adapters.py\", line 501, in send
  raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
}}"}
{"timestamp": "2023-07-12T13:35:42.468812Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:36:34.267288Z", "level": "INFO", "name": "pycti.api", "message": "Message (delivery_tag=1) NOT acknowledged (RequestException)"}
{"timestamp": "2023-07-12T13:36:34.310458Z", "level": "INFO", "name": "pycti.api", "message": "Message processed, thread terminated"}
{"timestamp": "2023-07-12T13:36:34.312264Z", "level": "INFO", "name": "pycti.api", "message": "Message (delivery_tag=1) rejected"}
{"timestamp": "2023-07-12T13:36:34.341594Z", "level": "INFO", "name": "pycti.api", "message": "Processing a new message (delivery_tag=2), launching a thread..."}
{"timestamp": "2023-07-12T13:36:34.348972Z", "level": "INFO", "name": "pycti.api", "message": "Importing a marking-definition (id: marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da)"}
{"timestamp": "2023-07-12T13:36:42.482657Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:37:01.304246Z", "level": "ERROR", "name": "pycti.api", "message": "Execution timeout, too many concurrent call on the same entities"}
{"timestamp": "2023-07-12T13:37:21.669208Z", "level": "INFO", "name": "pycti.api", "message": "Importing a marking-definition (id: marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da)"}                                                                                                                 {"timestamp": "2023-07-12T13:37:42.493599Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:37:50.676912Z", "level": "ERROR", "name": "pycti.api", "message": "Execution timeout, too many concurrent call on the same entities"}                                                                                                                                             {"timestamp": "2023-07-12T13:38:01.609505Z", "level": "INFO", "name": "pycti.api", "message": "Importing a marking-definition (id: marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da)"}                                                                                                                 {"timestamp": "2023-07-12T13:38:42.507318Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:39:42.517684Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:40:42.529331Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
[...]
{"timestamp": "2023-07-12T13:55:42.707326Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:56:42.840551Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:56:58.482222Z", "level": "ERROR", "name": "pycti.api", "message": "Unsupported entity type for merging"}                                                                                                                                                                         {"timestamp": "2023-07-12T13:57:10.786746Z", "level": "INFO", "name": "pycti.api", "message": "Importing a marking-definition (id: marking-definition--34098fce-860f-48ae-8e50-ebd3cc5e41da)"}                                                                                                                {"timestamp": "2023-07-12T13:57:42.851256Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:58:42.863333Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
{"timestamp": "2023-07-12T13:59:42.874718Z", "level": "INFO", "name": "pycti.api", "message": "Getting connectors ..."}
[...]
```

</details>

## Other Relevant Links

The same issue: #2967. Suggestions from there:

- Improve the overall responsivness of your elasticsearch.
- Limit the number of workers, having too much is not always the best solution
- Incorporate more sources of information to shuffle the creation

---

Edit: found a workaround

**Comment by debelyoo** [comment]
Hello, I'm facing the same problem.

### TL;DR

I have ~500k bundles to import.
I have three workers. All of them are in deadlock and do no progress.

Interestingly, even with just 1 worker, I still get the same problem. This speaks somewhat against the deadlock assumption and potentially hints at some other problem with not releasing a lock (?).

### What I tried

- Restart the workers. The come back to deadlock state immediately after restart.
- Restart the API. I still see the same lock errors (`too many concurrent call on the same entities`)
- Decrease the number of workers to 1. The single remaining worker is still in deadlock.
- Increase the number of workers. The new workers process some bundles for a while but they eventually get in deadlock state.
- More memory to Elasticsearch. 32 GB did not help.

### What helped
Increasing the number of workers is the only operation that helps processing the bundles in the queue. I started with 3, then 10, then 15 and so on. It seems that every time the nb of workers is increased the newly created workers can process some bundles.
However, this is counter-intuitive since increasing the nb of workers to solve `too many concurrent call on the same entities` error seems wrong.

**Comment by alice-debra** [comment]
Hello,
We wanted to follow up to check if you are still experiencing the problem. If we are unable to reproduce the bug within the next few days, we may need to close this ticket.
Please let us know at your earliest convenience. Thank you again for your valuable feedback. We truly appreciate your contribution in helping us improve.
Kind regards,
Alice

**Comment by debelyoo** [comment]
I'm still experiencing this kind of problems. 
Recently I reloaded many old events from a MISP server (with the MISP connector) and I had a lot of lock errors. I changed (up and down) the number of workers without finding a satisfying setup. The version of OpenCTI was 6.5.2.

**Comment by alice-debra** [comment]
Hello @debelyoo,
If you can try to update your platform, in 6.6.7 a related bug has been fixed (https://github.com/OpenCTI-Platform/opencti/issues/9582)
Otherwise, we'd have to see which entities are in error, if it's always the same ones?

**Comment by alice-debra** [comment]
Hello @debelyoo,
This error is not a problem, as the entity is finally able to get in and depends heavily on the data structure. For now, it's possible to reduce this a bit by improving the cluster performance, but it's not something you can definitely avoid.
Kind regards,
Alice
